src/python/pants/backend/jvm/subsystems/jvm_tool_mixin.py

```diff
@@ -145,6 +150,72 @@ def formulate_help():
     jvm_tool = cls.JvmTool(register.scope, key, classpath, main, custom_rules)
     JvmToolMixin._jvm_tools.append(jvm_tool)
 
+  # TODO: Deprecate .register_jvm_tool()?
+  class JvmToolDeclaration(datatype([
+      ('tool_name', text_type),
+      # If the tool's `main` class name is supplied the tool classpath will be shaded!
+      ('main', Exactly(text_type, type(None))),
+      ('classpath', tuple),
+      ('custom_rules', tuple),
+  ])):
+    """A mostly-typed specification for a JVM tool. Can be passed around across python modules."""
+
+    def __new__(cls, tool_name, main=None, classpath=None, custom_rules=None):
+      return super(JvmToolMixin.JvmToolDeclaration, cls).__new__(
+        cls, text_type(tool_name), main, tuple(classpath or []), tuple(custom_rules or []))
+
+  class JvmToolDeclError(Exception): pass
+
+  @classmethod
+  def register_jvm_tool_decl(cls, register, decl, **kwargs):
+    """Register a `JvmToolDeclaration` with `.register_jvm_tool()`."""
+    if not isinstance(decl, cls.JvmToolDeclaration):
+      raise cls.JvmToolDeclError(
+        'decl {} must be an instance of {}.{}'
+        .format(decl, cls.__name__, cls.JvmToolDeclaration.__name__))
+    kwargs = dict(
+      classpath=list(decl.classpath),
+      custom_rules=list(decl.custom_rules)
+    )
+    if decl.main:
+      kwargs['main'] = decl.main
+    cls.register_jvm_tool(register,
+                          decl.tool_name,
+                          **kwargs)
+
+  class CombinedJvmToolsError(JvmToolDeclError): pass
+
+  @classproperty
+  def combined_jvm_tool_names(cls):
+    """Return a list of tool names which can be invoked as a single jar.
+
+    This allows tools to be invoked one at a time, but with the combined classpath of all of them
+    using ensure_combined_jvm_tool_classpath(). This allows creating nailgun instances for the
+    tools which have the same fingerprint, and allows a task to invoke multiple different JVM tools
+    from the same nailgun instances. See #7089.
+    """
+    raise NotImplementedError('combined_jvm_tool_names is not implemented, see '
+                              'https://github.com/pantsbuild/pants/pull/7092')
+
+  @memoized_property
+  def _combined_jvm_tool_classpath(self):
+    cp = []
+    for tool_name in self.combined_jvm_tool_names:
+      cp.extend(self.tool_classpath(tool_name))
```

Comment:
To merge the classpath this way (ie, concatenation), I don't think the added `JvmToolDeclaration` here is necessary: a caller should be able to execute this same loop. If this works (and continues to work as the versions of these dependencies change) then I think you might consider having the caller just do this, as it's pretty straightforward.

Having said that: one reason to bake "merged tool support" into `JvmToolMixin` would be to _resolve_ a merged tool. Resolving a merged tool would mean doing an ivy or coursier resolve with all of the inputs at once, which would "sometimes" (how often in this case? unknown) result in different more compatible versions of some of the tool dependencies.

I think I'd lean toward the first thing.

Reply:
The reason why `JvmToolDeclaration` is here is so that we can use `ZINC_COMPILER_DECL` verbatim from `zinc.py` to avoid having the zinc we use in `RscCompile` fall out of sync with the `ZincCompile` zinc (just giving the kwargs to `register_jvm_tool()` a name). If that's not something we care too much about relative to the added complexity we can remove it.

Reply:
Resolving a merged tool sounds like the right idea (for a followup PR), but absolutely the right idea if we're planning to bake them into a single classpath anyway.

Reply:
> The reason why JvmToolDeclaration is here is so that we can use ZINC_COMPILER_DECL verbatim from zinc.py to avoid having the zinc we use in RscCompile fall out of sync with the ZincCompile zinc (just giving the kwargs to register_jvm_tool() a name). If that's not something we care too much about relative to the added complexity we can remove it.

`RscCompile` extends `ZincCompile`, so it should already have the zinc tools registered via the call to `super(..).register_options(..)`. So all of the tools should be accessible in the subclass by name.

> Resolving a merged tool sounds like the right idea (for a followup PR), but absolutely the right idea if we're planning to bake them into a single classpath anyway.

I'm not sure that we are planning to do that in the long term... this is currently a workaround for nailgun fingerprinting, which is subject to change.

Reply:
`ZincCompile` doesn't register zinc, though, the `Zinc` subsystem does. I could check to see how brittle it is and if moving the zinc compiler decl specifically into the `ZincCompile` task would still work.

Reply:
Removed all of this as per offline discussion!